### ISSUE TITLE
Remove Repetitive words

### ISSUE THREAD

**Issue by sreemoyk** [MaartenGr/BERTopic#639]
My dictionary consists of some words which are common in all the documents and present in large quantities. Hence, the topics generated constitutes of mostly these common words with a few others. Do we have a way to pre-process our data such that frequently occurring words are filtered out or not given much importance during modelling?

**Comment by MaartenGr** [comment]
A handy advantage of BERTopic is its use of the [`CountVectorizer`](https://maartengr.github.io/BERTopic/getting_started/countvectorizer/countvectorizer.html) is that you can process the documents *after* embedding them but *before* creating the topic representations. In other words, you can use the `stop_words` variable to remove any words you do not want ending up in the topic representations.

**Comment by MaartenGr** [comment]
Hmmm, just to be certain I understood correctly. There are specific words that you want to remove from all documents, correct? If so, then you could pass these words to the `stop_words` parameter as follows: `CountVectorizer(stop_words=["progress", "pupils"])`. That would remove those words across all documents.

**Comment by sreemoyk** [comment]
Yes your understanding is correct. However, since there are around 13k documents, there are many words that are occurring frequently. It is difficult to identify all those words. I tried using TF-IDF to compute the tfidf scores of words in a document and remove the words with low scores, but the I am getting only two topics after model execution. 
What do you suggest how should I approach this?

**Comment by MaartenGr** [comment]
> However, since there are around 13k documents, there are many words that are occurring frequently. It is difficult to identify all those words. 

Good question. There are several ways of approaching this but before going into that perhaps it would also be worthwhile if you could share the code for training BERTopic. There might be some optimization that could help. Also, how many topics do you get as an output? 

> What do you suggest how should I approach this?

There is something that you can try out. If there are some words that are incredibly frequent but generally do not contain informative information, you could use the `max_df` in the `CountVectorizer` instead:

**max_df : float in range [0.0, 1.0] or int, default=1.0**
`
When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words). If float, the parameter represents a proportion of documents, integer absolute counts. This parameter is ignored if vocabulary is not None.
`

In other words, setting this value to a high float between 0 and 1, for example, `0.9`, will remove words that appear in essentially all topics.

> I tried using TF-IDF to compute the tfidf scores of words in a document and remove the words with low scores, but the I am getting only two topics after model execution.

This is surprising to me as removing stop words with the `CountVectorizer` in BERTopic does not affect the number of topics at all. I definitely think it might be worthwhile to look a bit closer at the training procedure.

**Comment by sreemoyk** [comment]
You have pointed out an interesting parameter which I have tried in different variations as below-
1.  BerTopic with CV(with stopwords= english and 2 other words)
BERTopic(vectorizer_model=cv,verbose=True,calculate_probabilities=True)
Result 
![image](https://user-images.githubusercontent.com/94416690/180958237-d2055717-d8ec-48c7-8a31-8fdc8af41876.png)

2. BerTopic with CV(with stopwords= english and 2 other words, max_df=0.90)
![image](https://user-images.githubusercontent.com/94416690/180958717-397b4a6d-5cb2-4c18-bcf5-e969e050bb4b.png)

As you can see the first image consists of topics with similar names even though the corpus has 13k documents. The second one after using max_df the number of topics I get is just 3.

**Comment by MaartenGr** [comment]
The second results are indeed rather odd. Based on the output, it seems that the tokenizer works differently. Could you share the exact code for training the BERTopic model, including the code for creating the vectorizers? Also, did you preprocess the documents in some way?

**Comment by sreemoyk** [comment]
For the 2nd result:- 
stopwords = nltk.corpus.stopwords.words('english')
newStopWords = ['pupils','progress']
stopwords.extend(newStopWords)
CountVectorizer(stop_words=stopwords,max_df=0.90)
BERTopic(vectorizer_model=cv,verbose=True,calculate_probabilities=True)

No preprocessing apart from stopwords

**Comment by gillespied** [comment]
I have also noticed that changing the vectorizer can reduce the number of topics. The effect is not as dramatic as here. In this case it looks as if there maybe no words left to describe the topics, that is all words for topics 3 through 8 have a df over 0.90.

**Comment by sreemoyk** [comment]
@gillespied For instance 1, we can see that see that many topic names are similar and also contain similar words, is there a way to extract different topics? I have tried diversity parameter but that is responsible for within topic distribution and not on the overall corpus.

**Comment by gillespied** [comment]
@sreemoyk do the tests from any of your topics 3 through 8 contain words other than those in the nltk stopword list (+your two words)?

**Comment by sreemoyk** [comment]
@gillespied yes, there are many other words in the topics from 3 through 8. I am particularly confused of instance 2, where words in the topics are concatenated. @MaartenGr any idea on that?

**Comment by sreemoyk** [comment]
@MaartenGr I am able to get distinct topics now. I observed that splitting each document's content thereby increasing the number the rows did the trick for me. Thanks for your prompt responses.
